lib/membrane_audio_interleaver.ex

```diff
@@ -0,0 +1,289 @@
+                description: """
+                The value defines a raw audio format of pads connected to the
+                element. It should be the same for all the pads.
+                """,
+                default: nil
+              ],
+              frames_per_buffer: [
+                type: :integer,
+                spec: pos_integer(),
+                description: """
+                Assumed number of raw audio frames in each buffer.
+                Used when converting demand from buffers into bytes.
+                """,
+                default: 2048
+              ],
+              order: [
+                type: :list,
+                spec: [any()],
+                description: """
+                Order in which channels should be interleaved
+                """
+              ]
+
+  def_output_pad :output,
+    mode: :pull,
+    availability: :always,
+    caps: Caps
+
+  def_input_pad :input,
+    mode: :pull,
+    availability: :on_request,
+    demand_unit: :bytes,
+    caps: Caps
+
+  @impl true
+  def handle_init(%__MODULE__{} = options) do
+    state =
+      options
+      |> Map.from_struct()
+      |> Map.merge(%{
+        pads: %{},
+        finished: false,
+        channels: length(options.order)
+      })
+
+    {:ok, state}
+  end
+
+  @impl true
+  def handle_pad_added(pad, _context, state) do
+    state =
+      Bunch.Access.put_in(
+        state,
+        [:pads, pad],
+        %{queue: <<>>, stream_ended: false}
+      )
+
+    {:ok, state}
+  end
+
+  @impl true
+  def handle_pad_removed(pad, _context, state) do
+    state = Bunch.Access.delete_in(state, [:pads, pad])
+    {:ok, state}
+  end
+
+  @impl true
+  def handle_prepared_to_playing(
+        _context,
+        %{input_caps: %Caps{} = input_caps, channels: channels} = state
+      ) do
+    {{:ok, caps: {:output, %Caps{input_caps | channels: channels}}}, state}
+  end
+
+  @impl true
+  def handle_prepared_to_playing(_context, %{input_caps: nil} = state) do
+    {:ok, state}
+  end
+
+  @impl true
+  def handle_demand(:output, size, :bytes, _context, %{channels: channels} = state) do
+    do_handle_demand(div(size, channels), state)
+  end
+
+  @impl true
+  def handle_demand(
+        :output,
+        buffers_count,
+        :buffers,
+        _context,
+        %{frames_per_buffer: frames, input_caps: input_caps} = state
+      ) do
+    case input_caps do
+      nil ->
+        {:ok, state}
+
+      _caps ->
+        size = buffers_count * Caps.frames_to_bytes(frames, input_caps)
+
+        do_handle_demand(size, state)
+    end
+  end
+
+  @impl true
+  def handle_end_of_stream(pad, _context, %{input_caps: input_caps} = state) do
+    if state.finished do
+      # end of stream already sent
+      {:ok, state}
+    else
+      sample_size = Caps.sample_size(input_caps)
+
+      state =
+        case Bunch.Access.get_in(state, [:pads, pad]) do
+          %{queue: queue} when byte_size(queue) < sample_size ->
```

Comment:
so I shouldn't consider the possibility that I received the number of bytes that isn't the multiple of sample_size?

Reply:
Bear in mind this is at the end of the stream - if you received only part of the sample then something is broken - it should not happen. I guess if the queue is not empty you should warn about it - this might indicate the caps were wrong so you produced garbage or the input is corrupt and there were no parser to catch that

Reply:
Pasting suggested code I get an error "(CompileError) undefined variable ^pad. No variable "pad" has been defined before the current pattern". I tried it both with and without ^.